rse-py-package.Rmd

```diff
@@ -0,0 +1,1460 @@
+> This was...a good survival trait.
+> It meant that in times of danger you spent the day sitting very quietly in a building with very thick walls.
+>
+> -- Terry Pratchett
+
+The more software you write,
+the more you realize that a programming language is a way to build and combine software libraries.
+Every widely-used language now has an online repository
+from which people can download and install those libraries.
+This lesson shows you how to use Python's tools to create and share libraries of your own.
+
+This material is based in part on [Python 102][python-102] by [Ashwin Srinath][srinath-ashwin].
+
+## How can I turn a set of Python source files into a module? {#rse-py-package-modules}
+
+Any Python source file can be imported by any other.
+When a file is imported,
+the statements in it are executed as it loads.
+Variables defined in the file are then available as `module.name`.
+(This is why Python files should be named using [`pothole_case`](glossary.html#pothole-case)
+instead of [`kebab-case`](glossary.html#kebab-case):
+an expression like `import some-thing` isn't allowed
+because `some-thing` isn't a legal variable name.)
+
+As an example,
+we can put a constant and two functions used in our Zipf's Law study in a file called `zipf.py`:
+
+<!-- src="rse-py-package/01/zipf.py" -->
+```python
+from pytest import approx
+
+RELATIVE_ERROR = 0.05
+
+def make_zipf(length):
+    assert length > 0, 'Zipf distribution must have at least one element'
+    result = [1/(1 + i) for i in range(length)]
+    return result
+
+
+def is_zipf(hist, rel=RELATIVE_ERROR):
+    assert len(hist) > 0, 'Cannot test Zipfiness without data'
+    scaled = [h/hist[0] for h in hist]
+    perfect = make_zipf(len(hist))
+    return scaled == approx(perfect, rel=rel)
+```
+
+<!-- == noindent -->
+and then use `import zipf`,
+`from zipf import is_zipf`,
+and so on:
+
+<!-- src="rse-py-package/01/use.py" -->
+```python
+from zipf import make_zipf, is_zipf
+
+generated = make_zipf(5)
+print('generated distribution: {}'.format(generated))
+generated[-1] *= 2
+print('passes test with default tolerance: {}'.format(is_zipf(generated)))
+print('passes test with tolerance of 1.0: {}'.format(is_zipf(generated, rel=1.0)))
+```
+
+Running this program produces the following output:
+
+```text
+generated distribution: [1.0, 0.5, 0.3333333333333333, 0.25, 0.2]
+passes test with default tolerance: False
+passes test with tolerance of 1.0: True
+```
+
+<!-- == noindent -->
+It also creates a sub-directory called `__pycache__`
+that holds the compiled versions of the imported files.
+The next time Python imports `zipf`,
+it checks the timestamp on `zipf.py` and the timestamp on the corresponding file in `__pycache__`.
+If the latter is more recent,
+Python doesn't bother to recompile the file:
+it just loads the bytes in the cached version and uses those.
+To avoid confusing it,
+we (almost) always put `__pycache__` in `.gitignore`.
```

Comment:
almost? (maybe an exercise at the end asking when we wouldn't?)

Reply:
Added a bit of a rationale here